cpu/stm32_common/periph/uart.c

```diff
@@ -145,7 +145,10 @@ static inline void wait_for_tx_complete(uart_t uart)
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
     assert(uart < UART_NUMOF);
-
+    /* If tx is not enabled don't try to send */
+    if (!(dev(uart)->CR1 & USART_CR1_TE)) {
```

Comment:
I don't think this should be enabled always, but be an assertion.

Reply:
Well, `assert()` uses `stdio` as well. I'm not sure what would happen than. In any case, the user will not be able to "see" that the assertion failed, as no output over UART is available.

Reply:
mmm, I get the point. The problem is that having the UART fail silently will be very surprising for the user.

Reply:
I think the best user experience would be to make sure `stdio` becomes available as soon as possible. This would allow us using `DEBUG()` even in the initialization code of other stuff.

Maybe a good compromise would be to perform this check only when `DEVELHELP` is enabled. And also set some global flag (only present with `DEVELHELP`) when this occurs. At the very end of the initialization of `stdio` a check for that flag could be added (again, only with `DEVELHELP`) that would print a warning about lost `stdio` output.

Reply:
Then the current implementation would crash every-time you try to debug a periph.

Reply:
It currently calls uart_write before init_uart as pointed out by the issue.

Reply:
That's why I like error codes everywhere.  As it stands the periphs get initialized before the uart does and attempting to write on certain boards locks it up in an infinite loop waiting for a flag.  An assert would fail if added.  Maybe that is the behavior we desire but then all periphs that want to be debugged and have something in the init would crash.

Reply:
> I think the best user experience would be to make sure stdio becomes available as soon as possible. 

100%. This PR is kind of a work-around for me. The fix would be to either to split board initialization:
https://github.com/RIOT-OS/RIOT/blob/b709e63581a85a504b1020a025ef6f0e73f80b1d/cpu/cortexm_common/vectors_cortexm.c#L124-L133

into two functions that get called before and after the libc init, or to make board init responsible for initializing the C library.

In any case it is not an easy task, that's why I'm posting this as a comment and not a review, I don't want to block a PR that would prevent a lockup.

Reply:
> I think the best user experience would be to make sure `stdio` becomes available as soon as possible.

+1